scheduler/scheduler.py

```diff
@@ -2755,4 +2824,14 @@ def _done_callback(self, job_id, worker_id, all_num_steps,
             for single_job_id in to_remove:
                 self._remove_job(single_job_id)
 
+            # Re-dispatch jobs with extended leases that are still active
```

Comment:
Why do we need to do this?

Reply:
So the only time at which jobs are dispatched other than the initial round is halfway through each round. Let's say a job is running in round `r`, has its lease extended halfway through round `r`, but then completes before round `r` is finished (this might happen if previous rounds finished earlier than expected and the round completion times skew a bit). Then the scheduler will be waiting for the job to complete before it can complete round `r+1`, but this will never happen because the job was never active in round `r+1` to begin with. So to fix this we can check when a job completes whether it's supposed to still be running and if so we can dispatch it again.